test/e2e/basic/ping-pong.go

```diff
@@ -2,160 +2,168 @@ package e2e
 
 import (
 	"context"
-	"fmt"
 	"io"
 	"io/ioutil"
-	"os"
 	"strings"
-	"time"
 
 	"github.com/onsi/ginkgo"
 	"github.com/onsi/gomega"
 	srov1beta1 "github.com/openshift-psap/special-resource-operator/api/v1beta1"
-	"github.com/openshift-psap/special-resource-operator/pkg/warn"
 	"github.com/openshift-psap/special-resource-operator/test/framework"
 	"github.com/pkg/errors"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/tools/clientcmd"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-var clientSet kubernetes.Clientset
+const (
+	pingPongName      = "ping-pong"
+	pingPongChartPath = "../../../charts/example/ping-pong-0.0.1/ping-pong.yaml"
+)
 
-var _ = ginkgo.Describe("[basic][ping-pong] create and deploy ping-poing", func() {
+var _ = ginkgo.Describe("[basic][ping-pong] Test ping-pong", func() {
 
 	cs := framework.NewClientSet()
 	cl := framework.NewControllerRuntimeClient()
-	ginkgo.It("Can create and deploy ping-pong", func() {
-		ginkgo.By("Creating ping-pong #1")
-		specialResourceCreate(cs, cl, "../../../charts/example/ping-pong-0.0.1/ping-pong.yaml")
-		checkPingPong(cs, cl)
-		specialResourceDelete(cs, cl, "../../../charts/example/ping-pong-0.0.1/ping-pong.yaml")
+	clientSet, err := GetKubeClientSet()
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+	ginkgo.BeforeEach(func() {
+		ginkgo.By("[pre] Creating ping pong SpecialResource...")
+		err := pingPongCreate(cs, cl)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		ginkgo.By("[pre] Waiting ping pong pods to be ready...")
+		err = waitPingPongReady(clientSet)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	})
+
+	ginkgo.AfterEach(func() {
+		ginkgo.By("[post] Deleting ping-pong SpecialResource...")
+		err := pingPongDelete(cs, cl)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		err = waitPingPongDeleted(cl)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	})
+
+	ginkgo.Context("when installed", func() {
+		ginkgo.It("Check logs", func() {
+			err := checkPingPong(clientSet, cs, cl)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		})
 	})
 })
 
-func GetPodLogs(namespace string, podName string, containerName string, follow bool) (string, error) {
-	count := int64(100)
+func getPodLogs(clientSet *kubernetes.Clientset, namespace string, podName string) (string, error) {
+	count := int64(10)
 	podLogOptions := v1.PodLogOptions{
-		//		Container: containerName,
-		Follow:    follow,
+		Follow:    false,
 		TailLines: &count,
 	}
 
 	podLogRequest := clientSet.CoreV1().
 		Pods(namespace).
 		GetLogs(podName, &podLogOptions)
-	stream, err := podLogRequest.Stream(context.TODO())
+	stream, err := podLogRequest.Stream(context.Background())
 	if err != nil {
 		return "", err
 	}
 	defer stream.Close()
 
-	var message string
-
 	buf := make([]byte, 2000)
```

Comment:
2000 bytes sounds arbitrary, and this seems to always read from the top. Could we maybe read into a `bytes.Buffer` in a loop to make sure that we read all lines?

Reply:
This is due to the logic inside the pod. We are only interested in finding a pattern in the logs, and just once is enough. There is no need to read the whole log because it repeats every 2 seconds, so a subset must contain it, as it is quite small.